api/src/entities/Chemeq.py

```diff
@@ -1,29 +1,88 @@
+                "formula_matrix": [[1, 0, 0, 0, 1], [0, 1, 0, 0, 0], [0, 0, 1, 0, 0], [0, 0, 0, 1, 1]],
+                "stoichiometric_matrix": [[1, 0, 0, 0, -1]],
+            }
+        }
+
+    @validator("number_of_components")
+    def validate_number_of_components(cls, v):
+        if v < 1:
+            raise ValueError("Number of components must be non-zero.")
+        return v
+
+    @validator("component_ids")
+    def validate_component_ids(cls, v, values):
+        # TODO: Possible to validate each ID against table
+        if len(v) != values["number_of_components"]:
+            raise ValueError("Length of component_ids must match number_of_components.")
+        return v
+
+    @validator("number_of_reactions")
+    def validate_number_of_reactions(cls, v, values):
+        if v < 1:
+            raise ValueError("Number of reactions must be non-zero.")
+        if values["number_of_components"] - v < 1:
+            raise ValueError("Number of reactions must be greater than number_of_components + 1")
+        return v
+
+    @validator("feed_composition")
+    def validate_feed_composition(cls, v, values):
+        if len(v) != values["number_of_components"]:
+            raise ValueError("Length of feed_composition must match number_of_components.")
+        return v
 
-    def to_dict(self):
-        return asdict(self)
+    @validator("formula_matrix")
+    def validate_formula_matrix(cls, v, values):
+        number_of_columns = values["number_of_components"]
+        number_of_rows = values["number_of_components"] - values["number_of_reactions"]
+        row_count = 0
+        for row in v:
+            if not len(row) == number_of_columns:
+                raise ValueError(f"Formula matrix must have {number_of_columns} columns")
+            row_count += 1
+        if not row_count == number_of_rows:
+            raise ValueError(f"Formula matrix must have {number_of_rows} rows")
+        return v
 
-    @classmethod
-    def from_dict(cls, dict_) -> "Chemeq":
-        class_fields = {f.name for f in fields(cls)}
-        return Chemeq(**{k: v for k, v in dict_.items() if k in class_fields})
+    @validator("stoichiometric_matrix")
+    def validate_stoichiometric_matrix(cls, v, values):
+        number_of_columns = values["number_of_components"]
+        number_of_rows = values["number_of_reactions"]
+        row_count = 0
+        for row in v:
+            if not len(row) == number_of_columns:
+                raise ValueError(f"Stoichiometric matrix must have {number_of_columns} columns")
+            row_count += 1
+        if not row_count == number_of_rows:
+            raise ValueError(f"Stoichiometric matrix must have {number_of_rows} rows")
+        return v
 
-    def compute(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    def compute(self) -> Tuple[npt.NDArray[bytes], npt.NDArray[float], npt.NDArray[float], npt.NDArray[float]]:
```

Comment:
Instead of using a basic tuple here, we should consider NamedTuple.
Usually much more pleasant to work with.
https://docs.python.org/3/library/collections.html#collections.namedtuple

Reply:
Nice! Makes the code a lot cleaner.